File: PolynomialRegression.py
#Polynomial Regression

#Importing the three essential libraries
import numpy as np  #contains mathematical tools
import matplotlib.pyplot as plt  #imported sub-library to help plot nice charts
import pandas as pd  #best library to import & manage datasets

#importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')

#create matrix of independent/explanatory variables
X = dataset.iloc[:, 1:2].values
Y = dataset.iloc[:, 2].values

#fitting linear regression to dataset for comparison
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, Y)

#fitting polynomial regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 3)
X_poly = poly_reg.fit_transform(X)
lin_reg2 = LinearRegression()
lin_reg2.fit(X_poly, Y)

#visualising the linear regression results
plt.scatter(X, Y, color = "red")
plt.plot(X, lin_reg.predict(X), color = "blue")
plt.title("Truth or Bluff (Linear Regression)")
plt.xlabel("Position Level")
plt.ylabel("Salary")
plt.show

#visualising the polynomial regression results
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, Y, color = "red")
plt.plot(X_grid, lin_reg2.predict(poly_reg.fit_transform(X_grid)), color = "blue")
plt.title("Truth or Bluff (Polynomial Regression)")
plt.xlabel("Position Level")
plt.ylabel("Salary")
plt.show
#a degree was added to make the PolyReg better

#predicting new result with Linear Regression
lin_reg.predict(6.5)

#predicting a new result with Polynomial Regression
lin_reg2.predict(poly_reg.fit_transform(6.5))

#when comparing the two models, PolyReg > LinReg

File: PolynomialRegression.r
#Polynomial Regression

#Import dataset
dataset = read.csv("Position_Salaries.csv")
#we do not need the job title for this analysis
dataset <- dataset[2:3]

#small dataset and whatnot, no need to split into train/test sets
#feature scaling is handled

#fitting Linear Regression to the dataset (for comparison)
lin_reg <- lm(formula = Salary ~ ., 
             data = dataset)
summary(lin_reg)

#fitting the Polynomial Regression to the dataset
#build polynomial terms
dataset$Level2 <- dataset$Level^2
dataset$Level3 <- dataset$Level^3
dataset$Level4 <- dataset$Level^4
poly_reg <- lm(formula = Salary ~ .,
               data = dataset)
summary(poly_reg)

#visualizing the linear regression results
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), 
    colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, dataset)),
            colour = 'blue') + 
  ggtitle("Truth of Bluff (Linear Regression)") +
    xlab("Level") +
    ylab("Salary")
  

#visualizing the polynomial regression results
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), 
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, dataset)),
            colour = 'blue') + 
  ggtitle("Truth of Bluff (Polynomial Regression)") +
  xlab("Level") +
  ylab("Salary")

#predict a new result with LinReg model
y_pred <- predict(lin_reg, data.frame(Level = 6.5))

#predict a new result with PolyReg model
y_pred2 <- predict(poly_reg, data.frame(Level = 6.5, 
                                        Level2 = 6.5^2, 
                                        Level3 = 6.5^3, 
                                        Level4 = 6.5^4))